feat: add resolver for deleteGDPRUser mutation

sivakar12 opened the pull request:
ISSUES CLOSED: #75

sivakar12 committed:
feat: add resolver for deleteGDPRUser mutation

ISSUES CLOSED: #75

ojongerius commented on src/dataLayer/mongo/user.js:
```diff
@@ -66,3 +66,24 @@ export async function createUser(root, vars, ctx) {
+  const exists = await asyncErrorHandler(
+    doesExist(UserModel, { accountLinkId })
+  );
+  if (isEmpty(exists)) {
```
I believe this should throw instead. I'd suggest including the `accountLinkId` used too. Something like so:

```js
    throw new Error('There is no account with: accountLinkId ' + accountLinkId);
```

Throwing like this will return a 200 -fine for GraphQL, and add the message to the  "errors" array if it exist.

ojongerius commented on src/dataLayer/mongo/user.js:
```diff
@@ -66,3 +66,24 @@ export async function createUser(root, vars, ctx) {
+    };
+  }
+  await asyncErrorHandler(UserModel.deleteOne({ accountLinkId }));
+  return {
```
Don't think this has to return anything. If anything it should be the ID of the deleted user.

sivakar12 replied:
The mutation schema was already defined before I started working on this. There the return type was specified as "HTTPStatus!". Shall I make it a boolean so that it will be a 'true' is the user is deleted and false it the user does not exist?

ojongerius replied:
I'm not sure, if anything we should return the deleted user. The caller should make sure their call was successful by asserting there were no errors.

sivakar12 replied:
OK

ojongerius commented on src/graphql/resolvers/GDPRUser.js:
```diff
@@ -1,10 +1,11 @@
   Mutation: {
-    createUser
+    createUser,
+    deleteGDPRUser: deleteUser
```
I think this can just be `deleteUser` instead.

sivakar12 replied:
Then I will have to change the schema too, which was not defined by me. Is that all right?

ojongerius replied:
Sure!

ojongerius commented on src/dataLayer/mongo/user-dataLayer.test.js:
```diff
@@ -108,3 +109,26 @@ describe('getUser', () => {
+    const result = await createUser({}, {}, validContext);
+    const { accountLinkId } = result;
+    const response = await deleteUser({}, { accountLinkId }, validContext);
+    expect(response.statusCode).toEqual(204);
```
No need to test for this status code.

ojongerius commented on src/dataLayer/mongo/user-dataLayer.test.js:
```diff
@@ -108,3 +109,26 @@ describe('getUser', () => {
+    const { accountLinkId } = result;
+    const response = await deleteUser({}, { accountLinkId }, validContext);
+    expect(response.statusCode).toEqual(204);
+    expect(response.message).toBeTruthy();
```
Try to use snapshots instead, see the other tests and https://facebook.github.io/jest/docs/en/snapshot-testing.html for reference.

ojongerius commented on src/dataLayer/mongo/user-dataLayer.test.js:
```diff
@@ -108,3 +109,26 @@ describe('getUser', () => {
+      { accountLinkId: '0000' },
+      validContext
+    );
+    expect(response.statusCode).toBe(404);
```
See comments about snapshots and not testing status codes.

ojongerius reviewed (changes requested):
Nice work! See comments in my review.

ojongerius commented:
@Bouncey do we still need separate GDPR logic? Shall we just add the normal user mutations?

sivakar12 commented:
It looks like the GraphQL type HTTPStatus is used only for this particular mutation. Now you say the status is not needed. So shall I remove that altogether?

ojongerius commented:
Hey @sivakar12 I'll have another look tomorrow. I think HTTPStatus has no place in our schema and would support removing it.
Just a heads up this is a new project for us, and we are developing the schema based on the first use case we had. Sometimes it's a little harder to get changes through as we are figuring this out as we go 😄

sivakar12 committed:
fix: make changes to deleteUser code

Use snapshot testing, modify deleteGDPRUser mutation and change error responses

ISSUES CLOSED: #75

sivakar12 commented:
I have done the first set of changes you requested. 
I can remove the whole HTTPStatus part and merge GDPRUser with User when you say so

ojongerius commented:
@sivakar12 noticed there is a TODO left by @Bouncey at https://github.com/freeCodeCamp/open-api/blob/staging/src/graphql/typeDefs/GDPR-User/mutation.js#L1 could you please remove it?

ojongerius commented:
Nice work @sivakar12 . Let's wait for @Bouncey to weigh in. He added the HTTPStatus and created the GDPRUser type.

sivakar12 committed:
docs: remove TODO comment in GDPRUser mutations file

sivakar12 commented:
Just realized something! Shouldn't we check whether the account being deleted is that of the logged in user? Can anybody delete any others' accounts?

ojongerius commented on src/dataLayer/mongo/user.js:
```diff
@@ -66,3 +66,17 @@ export async function createUser(root, vars, ctx) {
+export async function deleteUser(root, vars, ctx) {
+  const { decoded } = verifyWebToken(ctx);
+  const { accountLinkId } = vars;
+  const loggedInId = decoded[namespace + 'accountLinkId'];
```
Assert that accountLinkId from vars matches loggedInId

ojongerius commented on src/dataLayer/mongo/user-dataLayer.test.js:
```diff
@@ -108,3 +109,28 @@ describe('getUser', () => {
   });
 });
+
+describe('deleteUser', () => {
```
add test to assert one user can't delete another user

sivakar12 committed:
fix: check whether the logged in user is the user to delete

Bouncey reviewed (commented):
I added the `GDPRUser` type so that we could expose only the fields we wanted on the User type, but still have the ability to conform to `GDPR` with this other type. 

The `HTTPStatus` type is maybe me not letting go of `REST` just yet. I wanted a way to communicate if mutations that could be considered `POST`, `PUT` and `DELETE` were successful or not. If `graphql` is already doing it for us then that is great! 

I am not married to anything, if it can be done better, that's awesome! No need to hang around for me, merge when you feel it is OK @ojongerius :+1:

ojongerius commented on src/dataLayer/mongo/user-dataLayer.test.js:
```diff
@@ -108,3 +110,40 @@ describe('getUser', () => {
+    expect(searchResult).toBe(null);
+    done();
+  });
+  it('should return with an error message when a user not in db', async done => {
```
could you please rephrase so that it does not mention the database? I'd like to future proof for moving some logic to other services. `should return an error  message for a non existing user` would do.

ojongerius commented on src/dataLayer/mongo/user-dataLayer.test.js:
```diff
@@ -108,3 +110,40 @@ describe('getUser', () => {
+    }
+    done();
+  });
+  it('should return with an error when deleting the user not logged in', async done => {
```
could you please rephrase so it's a little more clear what we are testing? Something like `should return an error message when trying to delete another user` or `should only allow deletion of own account` would do.

ojongerius commented on src/dataLayer/mongo/user.js:
```diff
@@ -66,3 +66,20 @@ export async function createUser(root, vars, ctx) {
+  const { accountLinkId } = vars;
+  const loggedInId = decoded[namespace + 'accountLinkId'];
+
+  if (loggedInId !== accountLinkId) {
```
🙌

ojongerius commented:
I only had comments about names of test. I'll merge this and make those changes myself.

Thanks for the contribution @sivakar12 !
